Label on rois

will-moore opened the pull request:
OK, I am done with fixing issues I'm aware of. To summarise a couple of the recent changes:

 - The `hFlip`, `vFlip`, `textRotation` attributes are now stored only on the ShapeManager itself, not on individual text Shapes. If you e.g. rotate the panel in the right panel, the `panel_model` is triggered to re-render and that sets the flip/rotation value of the shapeManager. So we don't need to edit the Shapes JSON values or the `panel.shapes` model.
 - This keeps the model cleaner and fixes issues with copy/paste of shapes between panels of different rotation
 - Other minor bugs/issues addressed
 - @Tom-TBT this has your recent commit included
 - NB: Adding `Label` Shapes from OMERO is working (missed it before) but we don't yet support the adding of Text from other shapes (to create separate Text Shapes in figure). I think we can leave this for now?

@Rdornier Do you have time to give this a test?

EDIT: I included https://github.com/Rdornier/omero-figure/pull/3/commits/957362358c1e0faeb5267c273ad3c8faab07745c which achieves the same fix as https://github.com/ome/omero-figure/pull/643 (since it's a small change and that PR would have now conflict with this). It also makes sense to test together.

EDIT: Another tweak to functionality: If you delete the "A" label in the inset panel, the "A" shape Text will get deleted from the ROI in the parent image: [fbaf293](https://github.com/Rdornier/omero-figure/pull/3/commits/fbaf29362b4527c67333dc791999f1e69f963aa7)

will-moore committed:
WIP trying text edit options

will-moore committed:
shapeEditorTest.html page working

will-moore committed:
Cleanup unused changes to shapes

will-moore committed:
More cleanup - Add new Text option to shapeEditorTest

will-moore committed:
More shape-editor cleanup

will-moore committed:
Fix display of drag-select area. Remove redundent this.clearSelectedShapes()

will-moore committed:
Fix drag-select intersect region for Text shapes

will-moore committed:
Update ROI dialog to work with new shape-manager

will-moore committed:
Fix colours when creating new Text shape

will-moore committed:
disable text UI controls when invalid

will-moore committed:
Add textAnchor edit support to shape-editor

will-moore committed:
Add support for editing textAnchor in ROI dialog

will-moore committed:
Fix creation and update of Inset Rectangle & Label

will-moore committed:
Improve formatting of JSON export

will-moore committed:
Cleanup console.logs

Tom-TBT committed:
refactor export script text shapes for tiff and pdf

will-moore committed:
Fix temp.destroy() after pasting shapes

will-moore committed:
Fix bugs when creating Inset on rotated panel

will-moore committed:
Handle panel vertical/horizontal flip when creating Inset

will-moore committed:
Reduce PADDING as suggested by Tom

will-moore committed:
flake8 fixes

Tom-TBT committed:
adjust exports to smaller text padding

will-moore committed:
textRotation is internal to shape_manager

will-moore committed:
hFlip and vFlip internal to shape_manager, not stored in Shape itself

will-moore committed:
Set fillColor and Opacity on new Inset text shapes

will-moore committed:
this.shapeManager.setText(Text); in render_toolbar if no shapes selected

This means that if you choose to create Text shape, once the shape is
created then shape_manager.getText() return 'Text' which corresponds to
the newly created shape.

will-moore committed:
Export script draws Text shapes on top of others

will-moore committed:
Rename shapeScalingFactor (unclear it's a percentage) to shapeScalingFraction (now a fraction)

will-moore commented:
NB: that last commit updates `shapeScalingFactor` to follow the pattern above with `hFlip, vFlip, textRotation` where these are not set on the Shape itself but remain properties of the Shape_manager - Also `shapeScalingFactor` is renamed `shapeScalingFraction` (float) since `shapeScalingFactor` was ambiguous whether this was a percentage or fraction.

NB: this change was prompted by https://github.com/ome/omero-figure/pull/643 which will need to be refactored accordingly.

will-moore committed:
Use shapeScalingFraction on all shapes

will-moore committed:
uncomment adding Shapes in shapeEditorTest.js

will-moore committed:
Deleting Inset label in Inset Panel also removes corresponding Text-Shape in parent

will-moore committed:
Fix adding Label shapes from OMERO

will-moore committed:
cleanup getParentShapeCoords

Rdornier commented:
Hi,

Thanks for the job you did here ; looks great.

> Another tweak to functionality: If you delete the "A" label in the inset panel, the "A" shape Text will get deleted from the ROI in the parent image

Good !

> Adding Label Shapes from OMERO is working (missed it before) but we don't yet support the adding of Text from other shapes (to create separate Text Shapes in figure).

I'm just wondering what is the `label` we mentioned. Is it the `Text` shape that we can create in iviewer ?

Jumping on that topic, I realize that the comment of the OMERO shape cannot be loaded anymore from figure (when loading ROIs from OMERO, only the shape can be added ; its comment is not loaded anymore). This is probably coming from the fact that you dropped the idea of linking text to shape in Figure. However, I found useful to add the comment as text. I don't know if the following option(s) could fit:
- adding a checkbox next to the `Add` button, to specify if the comment has to be added as text or not
- linking the text to the shape (as already done for insets)

This was previously done by automatically loading the comment as text, linking it to the corresponding shape but keeping it hidden until the user shows the text. But I think this approach is not fitting anymore with the new code. 

Minor bugs

1. When scaling the stroke width of the arrow, the resulting seems to not be rounded, which modifies the layout

<img width="882" height="582" alt="Capture d’écran 2025-12-17 092219" src="https://github.com/user-attachments/assets/e47c6f9b-302a-4324-8c03-6074626a08df" />

<img width="1253" height="392" alt="Capture d’écran 2025-12-17 091109" src="https://github.com/user-attachments/assets/ca116c34-0d5b-4b46-96d0-61a8da37d5da" />


2. When multiple text are overlapping, it looks a bit messy on the panels but the pdf is fine. I think there is nothing that we can do, as it is maybe a feature of Raphael shapes

<img width="917" height="462" alt="Capture d’écran 2025-12-17 095630" src="https://github.com/user-attachments/assets/8d6a0d27-db23-4a41-be60-ee4596c8b588" />

3. Select a text shape, remove the displayed text (i.e. empty string), select another shape. Then, the first text shape cannot be selected anymore but it is still present in the list of ROIs. Maybe we could try to delete the text shapes that have empty strings when the user saves and close the dialog. That will avoid confusion on the number of ROIs linked to the panel (displayed three, but actually can see only two)

<img width="867" height="407" alt="image" src="https://github.com/user-attachments/assets/4273a84d-eeb6-459a-8c67-ddb44b19d871" />

Apart from that, I think the rest looks good. I didn't see any issue with export pdf/tiff as far as I tested

will-moore committed:
Fix strokeWidth of new Arrow

Also explicitly set fillOpacity to 0 as this is a better default for other shapes too

will-moore commented:
Thanks @Rdornier - yes, I'd just noticed the same bug as you (1) - This is because I was using the zoomScalingFraction when creating the Arrow, so we get a funny float value for strokeWidth (instead of the chosen strokeWidth). Should be fixed by the last commit above.

For (2), I guess this needs for the Text background element to be move toFront() etc with the Text element itself. I'll have a look at it, but since we update the Z-index of shapes quite a bit while editing them, I'm not sure how easy it will be to reliably get the same result as in the PDF

I'll have a go at (3) too...

will-moore commented:
> I realize that the comment of the OMERO shape cannot be loaded anymore from figure (when loading ROIs from OMERO, only the shape can be added ; its comment is not loaded anymore). This is probably coming from the fact that you dropped the idea of linking text to shape in Figure. However, I found useful to add the comment as text. I don't know if the following option(s) could fit

Yes, I have removed that for now. I think it would be good to support, but needs a bit more thought and testing to get it right and I'm not sure we have time for that (if we want to get this released this week)? I suggest we add that in a follow-up release which shouldn't be much later - There's already a bunch of other open PRs that would justify a new release soon. Alternatively we just hold off releasing this until the new year?

Rdornier commented:
> Yes, I have removed that for now. I think it would be good to support, but needs a bit more thought and testing to get it right and I'm not sure we have time for that (if we want to get this released this week)? I suggest we add that in a follow-up release which shouldn't be much later 

Ok sounds good !

will-moore committed:
Keep text.bkgdRect() right behind text.element when sorting Shape z-index

will-moore committed:
Fix and simplify ROI toolbar colorpicker dropdowns

will-moore commented:
I looked at the overlay issue and updated so that the text and backgrounds stay "together", but the z-index is defined by the "sorting" which uses area, to make sure that small shapes are on top (and are always selectable):

<img width="655" height="455" alt="Screenshot 2025-12-17 at 12 58 35" src="https://github.com/user-attachments/assets/a0c11884-82ec-474e-bc4e-761fd1aaa579" />

So the order won't be preserved in PDF export.

will-moore committed:
Remove any Text shapes with empty text

will-moore committed:
Add 'Text' shape to docs/figure_file_format.rst

Rdornier commented:
Look good to me, thanks.
Do you think we could already merge ?

will-moore commented:
@Rdornier I think you could merge this PR into your branch now. That would give us a better idea of what `#600` looks like and we can do final checks on that before merging. 👍
